### DISCUSSION TITLE
Problem with TextEditor that doesn't change anything

### DISCUSSION THREAD

**Discussion by PFlorient** [clientIO/joint#2352] [Q&A] [answered]
### Introduction

Hello,

I have a problem with TextEditor.
I have icons with a custom svg. 
Attached is the svg code:

```
<svg overflow="visible" id="uuid-d563bbad-711e-482d-b96e-c2df1a9ba38d" xmlns="http://www.w3.org/2000/svg">
  <g class="rotatable">
    <g class="scalable">
      <path d="M57.55,17.93v47.22c0,.8-.44,1.45-.96,1.45H25.56c-.67-4.37-4.45-7.71-9.01-7.71s-8.34,3.34-9.01,7.71H1.95c-.53,0-.95-.66-.95-1.45V17.93c0-.8,.42-1.45,.95-1.45H56.59c.53,0,.96,.66,.96,1.45Z" fill="#b0edea" stroke="#131232" stroke-linecap="round" stroke-linejoin="round" stroke-width="2"/>
      <path d="M25.69,68.02c0,5.05-4.1,9.14-9.14,9.14s-9.13-4.09-9.13-9.14c0-.48,.04-.95,.12-1.41,.67-4.37,4.45-7.71,9.01-7.71s8.34,3.34,9.01,7.71c.09,.46,.13,.94,.13,1.41Z" fill="none" fill-opacity="0.0" stroke="#131232" stroke-linecap="round" stroke-linejoin="round" stroke-width="2"/>
      <path d="M86.36,68.39c0,5.04-4.09,9.13-9.13,9.13s-9.14-4.09-9.14-9.13,4.09-9.13,9.14-9.13,9.13,4.09,9.13,9.13Z" fill="none" fill-opacity="0.0" stroke="#131232" stroke-linecap="round" stroke-linejoin="round" stroke-width="2"/>
      <path d="M96.83,49.36l-12.98-24.5c-.26-.48-.76-.78-1.3-.78h-23.55c-.81,0-1.45,.67-1.45,1.47v39.44c0,.81,.64,1.47,1.45,1.47h9.04c.99-4.19,4.74-7.31,9.23-7.31s8.24,3.12,9.24,7.31h9.02c.81,0,1.47-.66,1.47-1.47v-14.93c0-.24-.06-.48-.17-.69Zm-13.19,.41h-16.08c-2.12,0-3.86-1.74-3.86-3.86v-13.48c0-2.12,1.74-3.86,3.86-3.86h9.04c1.43,0,2.75,.8,3.41,2.07l7.06,13.48c1.34,2.56-.53,5.64-3.42,5.64Zm13.34,10.22h-4.76v-5.1h4.76v5.1Z" fill="none" fill-opacity="0.0" stroke="#131232" stroke-linejoin="round" stroke-width="2"/><path d="M83.65,49.77h-16.08c-2.12,0-3.86-1.74-3.86-3.86v-13.48c0-2.12,1.74-3.86,3.86-3.86h9.04c1.43,0,2.75,.8,3.41,2.07l7.06,13.48c1.34,2.56-.53,5.64-3.42,5.64Z" fill="#b0edea" fill-rule="evenodd" stroke="#131232" stroke-linejoin="round" stroke-width="2"/>
      <text transform="translate(45.23 6.92)" font-family="ArialMT, Arial" font-size="8" class="text1">titre</text>
      <text transform="translate(32.89 90.05)" font-family="ArialMT, Arial" font-size="8" class="text2">Description</text>
      <text transform="translate(98.32 51.97)" font-family="ArialMT, Arial" font-size="8" class="text3">Qté</text>
    </g>
  </g>
</svg>
```

As you can see I have several <text> tags with classes to differentiate them.  But this doesn't change the problem. 
When I select a text from my icon, I do have the div containing my editor created, but it doesn't modify my text at all.
What's more, when I click on the text to trigger the TextEditor, my cursor is positioned at 0:0, a long way from the position of my icon. 

Have I misconfigured something?

### Steps to reproduce

the icon was created this way:  

```
 export class Truck extends shapes.basic.Generic {
    portLabelMarkup = [{
      tagName: 'text',
      selector: 'portLabel'
    }];
    defaults() {
      return util.defaultsDeep({
        markup: *svg above*,
        type: 'app.Truck',
        attrs: {
          root: {
            magnet: false
          },
          '.text1': { 'font-size': 18, },
          '.text2': { 'font-size': 18, },
          '.text3': { 'font-size': 18, },
        }
      }, shapes.basic.Generic.prototype.defaults);
    }
  }
```

### Does your question relate to JointJS or JointJS+. Select both if applicable.

JointJS, JointJS+

**Comment by kumilingus** [comment]
And how do you open the `TextEditor`?

**Comment by PFlorient** [comment]
In the code as follows (classic from the doc) :
```
paper.on('element:pointerdblclick', function (elementView, evt) {
        joint.ui.TextEditor.edit(evt.target, {
          cellView: elementView,
          textProperty: ['attrs', 'label', 'text'],
          annotationsProperty: ['attrs', 'label', 'annotations']
        });
      });
```
Otherwise, as a user, I double-click on the text contained in the icon. 
Moreover, double-clicking on a default library icon does not cause this effect.

**Reply by kumilingus** [reply, marked as answer]
You need to provide a correct `textProperty`. There is no element with selector `label` in your markup (as opposed to the [built-in shape]( https://github.com/clientIO/joint/blob/master/src/shapes/standard.mjs#L35)).

If are not using JointJS [selectors](https://resources.jointjs.com/docs/jointjs/v3.7/joint.html#dia.Cell.markup) you need to use CSS selectors (e.g. `class`).

```js
paper.on('element:pointerdblclick', function (elementView, evt) {
  const selector = '.text1'; // or use the proper selector based on `evt.target`.
  joint.ui.TextEditor.edit(evt.target, {
    cellView: elementView,
    textProperty: ['attrs', selector, 'text'],
  });
});
```

**Comment by PFlorient** [comment]
Thank you so much, this has indeed solved the problem of the text not changing. 
To do this, I added an `@selector="XXX"` to each text tag in my svg and repeated myself by doing a 
`const selector = evt.target.parentElement.attributes['joint-selector'].value`
(evt.target returns only the <tspan> tag generated without any selector)

However a problem persists, when I double-click on my text the `<div>` tag with the 'textarea-container' class is placed on the top left of my canvas (adding the positions corresponding to the text in my svg)
This gives a placement like this :
`<div class="textarea-container" style="left: 44.73px; top: 5.22004px;">`
Whereas in the case of a square by default I have more realistic positions in my canva: 
`<div class="textarea-container" style="left: 1071.13px; top: 1242.8px;">`

Added to this (although I think it's related) is the fact that it completely shifts my own positioning in my canva to be right in the center of the div.

Did I miss anything else?

Thank you in advance for your reply

**Reply by kumilingus** [reply]
You added the `@selector` attribute? Are you now using the [util.svg](https://resources.jointjs.com/docs/jointjs/v3.7/joint.html#util.svg) for the markup?

Perhaps you could share the updated code with us.

**Reply by PFlorient** [reply]
Well, after a few tests, and some changes in my code, and using util.svg when generating my markup, that solves the problem. 
I've changed the svg quite a bit, so it's not impossible that I've closed a tag incorrectly (which still makes it usable in jointjs). 

I've started using the `@selector` in the svg for the reason that, while browsing the various topics here and elsewhere on the net, I came across this solution. 

Thank you for your assistance.
